plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/DBUtils.java

```diff
@@ -761,84 +761,64 @@ public static Object getAttributeValue(
+        @NotNull Object[] row,
+        @Nullable int[] nestedIndexes
+    ) {
         if (attribute.isCustom()) {
-            return DBVUtils.executeExpression(((DBDAttributeBindingCustom)attribute).getEntityAttribute(), allAttributes, row);
+            return DBVUtils.executeExpression(((DBDAttributeBindingCustom) attribute).getEntityAttribute(), allAttributes, row);
         }
 
-        int depth = attribute.getLevel();
-        if (depth == 0) {
-            final int index = attribute.getOrdinalPosition();
-            if (index >= row.length) {
-                log.debug("Bad attribute '" + attribute.getName() + "' index: " + index + " is out of row values' bounds (" + row.length + ")");
-                return null;
-            } else {
-                if (nestedIndexes == null) {
-                    return row[index];
-                } else {
-                    if (attribute.getDataKind() != DBPDataKind.ARRAY) {
-                        // Sibling non-array attribute
-                        return DBDVoid.INSTANCE;
-                    }
-                    Object colValue = row[index];
-                    if (colValue instanceof DBDCollection) {
-                        if (((DBDCollection) colValue).getItemCount() <= nestedIndexes[0]) {
-                            // Not an error. This collection is shorter than sibling collection
-                            return DBDVoid.INSTANCE;
-                        }
-                        return ((DBDCollection) colValue).getItem(nestedIndexes[0]);
-                    } else {
-                        log.debug("Index specified for non-collection attribute");
-                    }
-                }
-            }
+        final int depth = attribute.getLevel();
+        final int index = attribute.getTopParent().getOrdinalPosition();
+
+        if (depth == 0 && attribute != attribute.getTopParent()) {
+            log.debug("Top-level attribute '" + attribute.getName() + "' has bad top-level parent: '" + attribute.getTopParent().getName() + "'");
+            return null;
         }
-        Object curValue = row[attribute.getTopParent().getOrdinalPosition()];
-        int indexNumber = 0;
+
+        if (index >= row.length) {
+            log.debug("Bad attribute '" + attribute.getName() + "' index: " + index + " is out of row values' bounds (" + row.length + ")");
+            return null;
+        }
+
+        Object curValue = row[index];
+        int curNestedIndex = 0;
 
         for (int i = 0; i < depth; i++) {
             if (curValue == null) {
                 break;
             }
-            DBDAttributeBinding attr = attribute.getParent(depth - i - 1);
-            assert attr != null;
-            try {
-                int nestedIndex = 0;
-                if (curValue instanceof DBDCollection) {
-                    nestedIndex = nestedIndexes == null ? 0 : nestedIndexes[indexNumber++];
+
+            final int nestedIndex;
```

Comment:
Maybe it would be better to create a variable outside the loop, and reset it to 0 here.

Reply:
I want to keep it `final` to minimize the total amount of assignments to it. I believe it improves visual code flow.